src/plugins/compressors/sz_threadsafe_plugin.cc

```diff
@@ -0,0 +1,424 @@
+    get(options, "sz_threadsafe:accelerate_pw_rel_compression", &threadsafe_params->accelerate_pw_rel_compression);
+    get(options, "sz_threadsafe:data_type", &threadsafe_params->dataType);
+    get(options, "sz_threadsafe:app", &app);
+    get(options, "sz_threadsafe:user_params", &user_params);
+#if PRESSIO_SZ_VERSION_GREATEREQ(2,1,11,1)
+{
+    unsigned int temp;
+    if(get(options, "sz_threadsafe:exafel:peaks_segs", &exafel_peaks_segs) == pressio_options_key_set) {
+      exafel_params.peaksSegs = static_cast<uint16_t*>(exafel_peaks_segs.data());
+    }
+    if(get(options, "sz_threadsafe:exafel:peaks_rows", &exafel_peaks_rows) == pressio_options_key_set) {
+      exafel_params.peaksRows = static_cast<uint16_t*>(exafel_peaks_rows.data());
+    }
+    if(get(options, "sz_threadsafe:exafel:peaks_cols", &exafel_peaks_cols) == pressio_options_key_set) {
+      exafel_params.peaksCols = static_cast<uint16_t*>(exafel_peaks_cols.data());
+    }
+    if(get(options, "sz_threadsafe:exafel:num_peaks", &temp) == pressio_options_key_set) {
+      exafel_params.numPeaks=static_cast<uint64_t>(temp);
+    }
+    if(get(options, "sz_threadsafe:exafel:calib_panel", &exafel_calibPanel) == pressio_options_key_set) {
+      exafel_params.calibPanel = static_cast<uint8_t*>(exafel_calibPanel.data());
+    }
+    get(options, "sz_threadsafe:exafel:tolerance", &exafel_params.tolerance);
+    if(get(options, "sz_threadsafe:exafel:bin_size", &temp) == pressio_options_key_set) {
+      exafel_params.binSize=static_cast<uint8_t>(temp);
+    }
+    if(get(options, "sz_threadsafe:exafel:sz_dim", &temp) == pressio_options_key_set) {
+      exafel_params.szDim=static_cast<uint8_t>(temp);
+    }
+    if(get(options, "sz_threadsafe:exafel:peak_size", &temp) == pressio_options_key_set) {
+      exafel_params.peakSize=static_cast<uint8_t>(temp);
+    }
+    if (app == "ExaFEL") {
+      user_params = &exafel_params;
+    }
+}
+#endif
+#ifdef HAVE_RANDOMACCESS
+    get(options, "sz_threadsafe:random_access", &threadsafe_params->randomAccess);
+#endif
+
+    return 0;
+  }
+
+  int compress_impl(const pressio_data *input, struct pressio_data* output) override {
+    size_t r1 = pressio_data_get_dimension(input, 0);
+    size_t r2 = pressio_data_get_dimension(input, 1);
+    size_t r3 = pressio_data_get_dimension(input, 2);
+    size_t r4 = pressio_data_get_dimension(input, 3);
+    size_t r5 = pressio_data_get_dimension(input, 4);
+    int status = SZ_NSCS;
+    size_t outsize = 0;
+
+    /*unsigned char* compressed_data = SZ_compress_customize(app.c_str(), user_params,
+        libpressio_type_to_sz_type(pressio_data_dtype(input)),
+        pressio_data_ptr(input, nullptr),
+        r5,
+        r4,
+        r3,
+        r2,
+        r1,
+        &outsize,
+        &status
+        );*/
+    unsigned char* compressed_data=SZ_compress_customize_threadsafe("SZ",&threadsafe_params,SZ_FLOAT,input->data(),
+		    r5,r4,r3,r2,r1,&outsize,&status);
```

Comment:
Don't hardcode the app string as `"SZ"`, allow the user to modify this value.

Likewise, don't hardcode the input type, get it from input.

finally, you can't just pass threadsafe_params here, you need to handle the case where the user has provided a "userdata" value to use with appmode==exafel.

Reply:
In order to not hardcode the string "SZ", I would have to change compress_impl to send in a variable with input from the user about which compressor they wanted to use, correct?

Reply:
We'll do it the same way that the sz_plugin does it now.